bitbots_ball_filter/bitbots_ball_filter/ball_filter.py

```diff
@@ -109,83 +119,97 @@ def _dynamic_reconfigure_callback(self, config):
         self.filter_timer = self.create_timer(self.filter_time_step, self.filter_step)
         return SetParametersResult(successful=True)
 
-    def ball_callback(self, msg: PoseWithCertaintyArray, optional_distance=True):
-        """handles incoming ball messages"""
-        if msg.poses:
-            balls = sorted(msg.poses, reverse=True, key=lambda ball: ball.confidence)  # Sort all balls by confidence
-            ball = balls[0]  # Ball with highest confidence
-
-            if ball.confidence < self.min_ball_confidence:
-                return
-            self.last_ball_msg = ball
-            ball_buffer = PointStamped()
-            ball_buffer.header = msg.header
-            ball_buffer.point = ball.pose.pose.position
-            try:
-                self.ball = self.tf_buffer.transform(ball_buffer, self.filter_frame, timeout=rclpy.duration.Duration(seconds=0.3))
-                self.ball_header = msg.header
-            except (tf2.ConnectivityException, tf2.LookupException, tf2.ExtrapolationException) as e:
-                self.get_logger().warning(str(e))
-
-    def reset_filter_cb(self, req, response):
-        self.get_logger().info("Resetting bitbots ball filter...")
+    def reset_filter_cb(self, req, response) -> Tuple[bool, str]:
+        self.logger.info("Resetting bitbots ball filter...")
         self.filter_initialized = False
-        return True, ""
+        response.success = True
+        return response
+
+    def ball_callback(self, msg: BallArray) -> None:
+        if msg.balls:  # Balls exist
+            # Either select ball closest to previous prediction or with highest confidence
+            if self.closest_distance_match:  # Select ball closest to previous prediction
+                ball_msg = self._get_closest_ball_to_previous_prediction(msg)
+            else:  # Select ball with highest confidence
+                ball_msg = sorted(msg.balls, key=lambda ball: ball.confidence.confidence)[-1]
+            self.ball = BallWrapper(self._get_transform(msg.header, ball_msg.center), msg.header, ball_msg.confidence.confidence)
+
+    def _get_closest_ball_to_previous_prediction(self, ball_array: BallArray) -> Union[Ball, None]:
+        closest_distance = math.inf
+        closest_ball_msg = ball_array.balls[0]
+        for ball_msg in ball_array.balls:
+            ball_transform = self._get_transform(ball_array.header, ball_msg.center)
```

Comment:
You should maybe do an optimization. Normally we have only a few balls, so this is not that significant, but the transform of the forward kinimatic is queried for each ball using the header of the array. This transform is constant across a given array, so you can query the transform for the array at the beginning and apply it for all balls in the loop.